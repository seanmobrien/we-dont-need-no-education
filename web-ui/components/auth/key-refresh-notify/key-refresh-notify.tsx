'use client';

import { useEffect, useState } from 'react';
import { useNotifications } from '@toolpad/core';
import { useSession } from '../session-provider';

const NOTIFICATION_KEY_SESSION_LOADING = 'session-loading';
const NOTIFICATION_KEY_KEY_VALIDATION = 'key-validation-status';

export const KeyRefreshNotify = () => {
  const notifications = useNotifications();
  const { status: sessionStatus, keyValidation } = useSession();
  
  // Track notification state to prevent duplicate notifications
  const [lastNotifiedSessionStatus, setLastNotifiedSessionStatus] = useState<string>('');
  const [lastNotifiedKeyStatus, setLastNotifiedKeyStatus] = useState<string>('');
  const [hasShownInitialLoadingToast, setHasShownInitialLoadingToast] = useState(false);

  // Handle session loading notification (only on initial mount)
  useEffect(() => {
    if (typeof window === 'undefined') return;

    if (sessionStatus === 'loading' && !hasShownInitialLoadingToast) {
      notifications.show('Loading session details...', {
        severity: 'info',
        key: NOTIFICATION_KEY_SESSION_LOADING,
        // No auto-timeout - will be closed when session loads
      });
      setHasShownInitialLoadingToast(true);
      setLastNotifiedSessionStatus(sessionStatus);
    } else if (sessionStatus !== 'loading' && lastNotifiedSessionStatus === 'loading') {
      // Close loading notification when session is no longer loading
      notifications.close(NOTIFICATION_KEY_SESSION_LOADING);
      setLastNotifiedSessionStatus(sessionStatus);
    }
  }, [sessionStatus, notifications, hasShownInitialLoadingToast, lastNotifiedSessionStatus]);

  // Handle key validation notifications
  useEffect(() => {
    if (typeof window === 'undefined') return;

    const { status: keyStatus, error, lastValidated } = keyValidation;
    const currentStatusKey = `${keyStatus}-${error || ''}-${lastValidated?.getTime() || 0}`;

    // Skip if we've already notified for this exact state
    if (currentStatusKey === lastNotifiedKeyStatus) {
      return;
    }

    switch (keyStatus) {
      case 'unknown':
        // Close any existing notifications but don't show new ones
        notifications.close(NOTIFICATION_KEY_KEY_VALIDATION);
        break;

      case 'invalid':
        // Keys are out of date, show info notification with no timeout
        notifications.show('Authentication keys are out of date and being refreshed...', {
          severity: 'info',
          key: NOTIFICATION_KEY_KEY_VALIDATION,
          // No autoHideDuration - will be replaced by success/error
        });
        break;

      case 'validating':
      case 'synchronizing':
        // Keys are being processed, show info notification with no timeout
        notifications.show('Refreshing authentication keys...', {
          severity: 'info',
          key: NOTIFICATION_KEY_KEY_VALIDATION,
          // No autoHideDuration - will be replaced by success/error
        });
        break;

      case 'failed':
        // Show error notification with custom timeout
        const errorMessage = error || 'Key validation failed. Please try again later.';
        notifications.show(errorMessage, {
          severity: 'error',
          key: NOTIFICATION_KEY_KEY_VALIDATION,
          autoHideDuration: 60000, // 60 seconds
        });
        break;

      case 'valid':
        // Keys are valid, close any existing notifications (no success message needed)
        notifications.close(NOTIFICATION_KEY_KEY_VALIDATION);
        break;

      case 'synchronized':
        // Keys were successfully synchronized, show success message
        notifications.show('Authentication keys have been successfully refreshed.', {
          severity: 'success',
          key: NOTIFICATION_KEY_KEY_VALIDATION,
          autoHideDuration: 7500, // 7.5 seconds
        });
        break;

      default:
        // Unknown status, close notifications
        notifications.close(NOTIFICATION_KEY_KEY_VALIDATION);
        break;
    }

    setLastNotifiedKeyStatus(currentStatusKey);
  }, [keyValidation, notifications, lastNotifiedKeyStatus]);

  return null;
};