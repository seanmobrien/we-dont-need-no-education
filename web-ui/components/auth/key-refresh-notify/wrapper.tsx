'use client';

import { NotificationsProvider } from "@toolpad/core";
import { KeyRefreshNotify } from './key-refresh-notify';

export const KeyRefreshNotifyWrapper = () => {
  return (
    <NotificationsProvider>
      <KeyRefreshNotify />
    </NotificationsProvider>
  );
};