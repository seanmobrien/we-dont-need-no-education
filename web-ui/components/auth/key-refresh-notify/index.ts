import { KeyRefreshNotifyWrapper as KeyRefreshNotify } from './wrapper';
export { KeyRefreshNotify };
