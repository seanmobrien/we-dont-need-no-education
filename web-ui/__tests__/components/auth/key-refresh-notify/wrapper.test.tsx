import React from 'react';
import { render } from '@/__tests__/test-utils';
import { KeyRefreshNotifyWrapper as KeyRefreshWrapper } from '@/components/auth/key-refresh-notify/wrapper';

describe('KeyRefreshWrapper', () => {
  it('renders with default state snapshot', () => {
    const { container } = render(
      <KeyRefreshWrapper>
        <div>Test child content</div>
      </KeyRefreshWrapper>
    );
    expect(container).toMatchSnapshot();
  });

  it('renders notification UI', () => {
    const { getByText } = render(
      <KeyRefreshWrapper>
        <div>Test child content</div>
      </KeyRefreshWrapper>
    );
    
    expect(getByText('Loading session details...')).toBeInTheDocument();
  });
});