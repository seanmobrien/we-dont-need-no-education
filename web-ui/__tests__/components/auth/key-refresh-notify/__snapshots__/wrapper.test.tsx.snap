// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`KeyRefreshWrapper renders with default state snapshot 1`] = `
<div>
  <div>
    Test child content
  </div>
  <div
    class="MuiSnackbar-root MuiSnackbar-anchorOriginBottomLeft css-1w76183-MuiSnackbar-root"
    role="presentation"
  >
    <span
      class="MuiBadge-root css-13xwjag-MuiBadge-root"
      direction="up"
      style="opacity: 1; transform: scale(1, 1); transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,transform 150ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
    >
      <div
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation0 MuiAlert-root MuiAlert-colorInfo MuiAlert-standardInfo MuiAlert-standard css-vwet49-MuiPaper-root-MuiAlert-root"
        role="alert"
        style="--Paper-shadow: var(--mui-shadows-0); --Paper-overlay: var(--mui-overlays-0);"
      >
        <div
          class="MuiAlert-icon css-vab54s-MuiAlert-icon"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-iilnq5-MuiSvgIcon-root"
            data-testid="InfoOutlinedIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M11,9H13V7H11M12,20C7.59,20 4,16.41 4,12C4,7.59 7.59,4 12,4C16.41,4 20,7.59 20, 12C20,16.41 16.41,20 12,20M12,2A10,10 0 0,0 2,12A10,10 0 0,0 12,22A10,10 0 0,0 22,12A10, 10 0 0,0 12,2M11,17H13V11H11V17Z"
            />
          </svg>
        </div>
        <div
          class="MuiAlert-message css-zioonp-MuiAlert-message"
        >
          Loading session details...
        </div>
        <div
          class="MuiAlert-action css-rgppqo-MuiAlert-action"
        >
          <button
            aria-label="Close"
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-sizeSmall css-11wq6a3-MuiButtonBase-root-MuiIconButton-root"
            tabindex="0"
            title="Close"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-e0e2zh-MuiSvgIcon-root"
              data-testid="CloseIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
              />
            </svg>
          </button>
        </div>
      </div>
      <span
        class="MuiBadge-badge MuiBadge-standard MuiBadge-invisible MuiBadge-anchorOriginTopRight MuiBadge-anchorOriginTopRightRectangular MuiBadge-overlapRectangular MuiBadge-colorPrimary css-7pto8m-MuiBadge-badge"
      />
    </span>
  </div>
</div>
`;
